### PR TITLE
Refactor: use standard string prototype methods in macros

### PR DESCRIPTION
**Pull request by bershanskiy**
Kumascript has a bunch of methods for operations on strings, which are exposed as the `string` global: https://github.com/mdn/yari/blob/master/kumascript/src/api/string.js

Since most of them are just wrappers around the standard string prototype, they look confusing (at least to me). This commit just strips these unnecessary wrappers. This does not change any behavior.

Tested with CI and by building `mdn/content@2d423bd` (the `HEAD` at the time of writing) and observing the same counts of flaws as before this commit.

This commit does not remove usage of `string.asyncReplace` (I'll file a follow-up for that).
This commit does not remove `kumascript/src/api/string.js` itself just in case someone plans to add more legacy macros to the repo.

### CODE REVIEW

File: kumascript/macros/WebkitBug.ejs
Comment (bershanskiy) [inline kumascript/macros/WebkitBug.ejs:13]: I deleted this comment because [referenced bug](https://bugzilla.mozilla.org/show_bug.cgi?id=765642#c7) was closed as WONTFIX.

File: kumascript/macros/Deprecated_Inline.ejs
Comment (bershanskiy) [inline kumascript/macros/Deprecated_Inline.ejs:20]: No need to check for `$0 != undefined`, since `$0` being truthy implies `$0 != undefined`.

File: kumascript/macros/DiscussionList.ejs
Comment (bershanskiy) [inline kumascript/macros/DiscussionList.ejs:163]: Check for `param` being truthy might be unnecessary, I added it just for peace of mind.

**Comment by mattiapontonio** [comment]
I didn't reviewed all but I like the more familiar `JSON.parse` than a custom _maybe not documented_ function.

**Review by mattiapontonio — APPROVED** [review]
Approved the approach but didn't tested, just readed all.
Notes: 
- if unused the old functions should be deleted.
- it's a big PR

File: kumascript/macros/DiscussionList.ejs
Comment (mattiapontonio) [reply kumascript/macros/DiscussionList.ejs:163]: That's a scope issue but You can add a default to the function.
No need to do It now.

**Review by peterbe — CHANGES_REQUESTED** [review]
I don't know why yet but the first time I noticed was that the sidebar is different in your branch vs `master` on 
http://localhost:5000/en-US/docs/Web/API/DataTransferItem/getAsFileSystemHandle
I.e. `Document: drag` vs. `HTMLElement: drag`

**Comment by peterbe** [comment]
Testing this was pretty hard. What I did was that I ran a complete build, on `master` and stuffed all the built `index.json` files in a temp directory. Then I did the same with your branch. 
Because plain `diff` can't cope since these .json files is just 1 single line, I ran:
```
fd index.json | xargs prettier --write
```
on each and every file. Now, when I ran `diff -r ` on the two directories, I noticed a couple where the `sidebarHTML` is different. Perhaps it needs a bit more careful scrutiny.

**Comment by bershanskiy** [comment]
@peterbe I'll go back to the drawing board then. Unfortunately, I don't have the time to prepare a fix today.

While preparing this PR, built everything before and after my patch to compare and after light checking only found differences in URL to GitHub which contained template with different ["Report Started"](https://github.com/mdn/yari/blob/85d4e28ebddab9fdb9a9643f23755fcf19a6dc42/client/src/document/on-github.tsx#L65). I'll just temporarily remove that and see what actual content changes.

**Comment by mattiapontonio** [comment]
In another project I versioned the `dist` so I saw the diffs in the PR, so pre PR a `build` and 1 commit, _like with package-lock.json_

**Comment by peterbe** [comment]
> @peterbe I'll go back to the drawing board then. Unfortunately, I don't have the time to prepare a fix today.
> 
> While preparing this PR, built everything before and after my patch to compare and after light checking only found differences in URL to GitHub which contained template with different ["Report Started"](https://github.com/mdn/yari/blob/85d4e28ebddab9fdb9a9643f23755fcf19a6dc42/client/src/document/on-github.tsx#L65). I'll just temporarily remove that and see what actual content changes.

If you want to, you can take this PR as a "template" and do many little PRs.

**Comment by bershanskiy** [comment]
I refined the testing strategy a bit. I created a small patch to make content builds "reproducible". With it, `diff -r` does not find any differences between builds before and after this PR. If you have time, would you be interested in trying this strategy on your system? It does take a lot of time, since it requires building content twice.

```bash
# Checkout latest mdn/yari master, ensure that branch is "clean" and deps are good
git checkout master
git status
yarn 
# Apply the commit below to make builds reproducible
git apply ../reproducible-builds.patch
# Build content, move it to "before" folder
yarn prepare-build
yarn build
mv client/build/ ../before

# Checkout PR branch, "clean" it, ensure deps are good
git checkout no-string-util
git checkout client/ kumascript/
git status
yarn
# Apply the commit below on top of "clean" PR branch
git apply ../reproducible-builds.patch
# Build content, copy it to "after" folder
yarn prepare-build
yarn build
mv client/build/ ../after
cd ..

# Make sure diff completes without finding any differences
diff -r before/ after/
```

The contents of `reproducible-builds.patch` from above:

```patch
diff --git a/client/src/document/on-github.tsx b/client/src/document/on-github.tsx
index c5e0142ce..1636f82b9 100644
--- a/client/src/document/on-github.tsx
+++ b/client/src/document/on-github.tsx
@@ -62,7 +62,6 @@ MDN URL: https://developer.mozilla.org$PATHNAME
 * Folder: \`$FOLDER\`
 * MDN URL: https://developer.mozilla.org$PATHNAME
 * GitHub URL: $GITHUB_URL
-* Report started: $DATE
 
 </details>
   `.trim();
diff --git a/kumascript/macros/LearnBox.ejs b/kumascript/macros/LearnBox.ejs
index 96ee7ce97..079e7e10f 100644
--- a/kumascript/macros/LearnBox.ejs
+++ b/kumascript/macros/LearnBox.ejs
@@ -74,7 +74,7 @@ for(let root of sources) {
 // DISPLAY
 // ----------------------------------------------------------------------------
 if (articles.length > 0) {
-    var a = articles[Math.floor(Math.random() * (articles.length))];
+    var a = articles[Math.floor(0.5 * (articles.length))];
 %>
 <div class="moreinfo">
   <p style="margin: 0 0 0.5em; font-size:120%;"><%- title %></p>
diff --git a/kumascript/macros/TenthCampaignQuote.ejs b/kumascript/macros/TenthCampaignQuote.ejs
index c237637cc..535d4543f 100644
--- a/kumascript/macros/TenthCampaignQuote.ejs
+++ b/kumascript/macros/TenthCampaignQuote.ejs
@@ -29,7 +29,7 @@ var index = 0;
 if (arguments.length > 0) {
     index = $0;
 } else {
-    index = Math.floor(Math.random() * quoteList.length);
+    index = Math.floor(0.5 * quoteList.length);
 }
 
 if (index > quoteList.length-1) {
```

**Comment by peterbe** [comment]
I trust your testing. The fact that you were able to so clearly explain your steps indicates that you know what you're doing. 
What I'll do is I'll just spot-check the sub-trees I found in my initial hacky diff'ing and see that they produce the same as before. 

However, @tannerdolby actually started [a PR](https://github.com/mdn/yari/pull/2073) before yours and I know it's going to clash. I think you're very good at this Yari and git stuff that you can probably resolve the conflicts once his things lands. Would be mind holding off a little bit and do a couple of merge conflicts (please merge in the latest upstream `master` rather than `rebase`)?

**Comment by peterbe** [comment]
@bershanskiy Only two macros that needs to resolved now.

**Comment by bershanskiy** [comment]
@peterbe Done!

I merged upstream (and cleaned up a couple lines) and tested with the [aforementioned method](https://github.com/mdn/yari/pull/2423#issuecomment-759003791) and it passed. I think, this is ready for review now.

**Comment by mattiapontonio** [comment]
Can this thing of testing if the build changes between branches be put into the CI in some refactoring cases like this? 🤔

**Comment by peterbe** [comment]
> I don't know why yet but the first time I noticed was that the sidebar is different in your branch vs `master` on
> http://localhost:5000/en-US/docs/Web/API/DataTransferItem/getAsFileSystemHandle
> I.e. `Document: drag` vs. `HTMLElement: drag`

http://localhost:3000/en-US/docs/Web/API/DataTransferItem/getAsFileSystemHandle is fixed now 🥳

File: kumascript/macros/PreviousMenuNext.ejs
Comment (peterbe) [inline kumascript/macros/PreviousMenuNext.ejs:45]: Wait, does this even work in 12? 
According to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll#browser_compatibility it was added in Node 15. 
I have Node 15 on my computer but our minimum is Node 12. 

The version of Node we use in CI is Node 12. If Node 12 doesn't support it, how come the tests didn't break?!

Comment (peterbe) [reply kumascript/macros/PreviousMenuNext.ejs:45]: Actually, what might have happened, in CI, is that the page is rendered but it has flaws (because of this), but that doesn't break the build. 

To be on the safe side, can we just replace (no pun intended) these with `$0.replace(/ /g, '_')`

File: kumascript/macros/geckoRelease.ejs
Comment (peterbe) [inline kumascript/macros/geckoRelease.ejs:5]: Feel free to use
```suggestion
var arg = `$0`.trim();
```
which is the more common "modern" way of doing it. Incidentally, in Babel, it gets converted to ES5 as basially what you had.

File: kumascript/macros/geckoRelease.ejs
Comment (peterbe) [inline kumascript/macros/geckoRelease.ejs:83]: It's not your fault but this is horrible :)
You look at the if statement above and based on that you think `geckoRelease` must be a string because the if statement checks if it's not a particular string. But strings don't have a `.join()`.

File: kumascript/macros/英語版章題.ejs
Comment (peterbe) [inline kumascript/macros/英語版章題.ejs:2]: ditto.

**Review by peterbe — CHANGES_REQUESTED** [review]
I think the CI tests needs to be fixed so that if you accidentally introduce something in the .ejs files, it should break. 

Can you please take care of the `.replaceAll` and port them back to the ancient way? 
In the meantime, I'll take a look at how the Kitchensink page works in CI.

File: kumascript/macros/geckoRelease.ejs
Comment (bershanskiy) [reply kumascript/macros/geckoRelease.ejs:83]: I don't like it either, so I refactored it a bit. Does it look better now?

File: kumascript/macros/geckoRelease.ejs
Comment (bershanskiy) [reply kumascript/macros/geckoRelease.ejs]: I'm not sure if <code>\`$0\`</code> will actually resolve `$0` or treat it as string `'$0'`. To avoid ambiguity, I used <code>\`${$0}\`</code>.

**Comment by bershanskiy** [comment]
@peterbe Thanks for the review!

I resolved merge conflicts and addressed your comments. Also, I tested once more using the test from above.

**Review by peterbe — APPROVED** [review]

**Pull request merged**
